### PR TITLE
Bug 1521157 - Update dataset for unit tests

### PR DESCRIPTION
**Pull request by camd**
Per a chat with Joel, here is the initial take on data we would display for failed "Tests" metric.  I'm sure this will continue to evolve.

This also adds a menu item to the push action menu to access this view.  I went back and forth on whether to include this.  Open to opinions if you think it's better NOT to have the menu item included at this time:
screenshot 2019-01-25 10 48 24

Here's the new look of the page:
screenshot 2019-01-25 10 50 28

### CODE REVIEW

**Comment by camd** [comment]
> would Ludicrous Speed imply a value=11?
> 
> My only other feedback on this is to help define what the numeric values of confidence mean- one could infer a 10 is the best and a 1 is the worse- is a 3 ok? an 8? etc.?

Agreed, yeah.  As you say, I was thinking 1-10, 10 being total confidence.  But it's pretty arbitrary at this point.  I think that's some calculus we'll have to come up with.  For the individual lines, do they average out to make the confidence for *Tests* overall?  Or does it take "the worst line" as the overall value?

"Confidence" may or may not be the right word, too.  Perhaps we set it to pass or fail and then the confidence is how much confidence we have in that assessment.  Or perhaps that's what you meant all along. :)

**Comment by jmaher** [comment]
In my mind our health score/test would be like:
7-10 - accept annotation (maybe some edge cases based on classification)
<7 = make it easy to retrigger and collect more data

we could do this further and say:
1-4 = no confidence at all, or no idea what this is
5-7 = a good chance this is a [infra, leak, intermittent, regression, etc.], but we need more data

of course retriggering a job 2 more times so we have 3 data points will be a good first step towards marking it as intermittent/noproblem if 2/3 are green.   There is an edge case here when a job has the opportunity for many frequent intermittent failures and run 1 fails on test_a.html, run 2 fails on test_b.html and run 3 fails on test_c.html - therefore it could look like perma fail- since we are doing test level failures instead of job/task level.

Of course, if we have health score in 8-10 range  for a regression, then we might want to avoid showing a retrigger button until they fix the regression and push again.

File: ui/push-health/Metric.jsx [hunk 1]
```diff
@@ -25,29 +27,56 @@ export default class Metric extends React.PureComponent {
...
     return (
       <td>
         <div className="d-flex flex-row">
           <div className={`bg-${resultColor} pr-2 mr-2`} />
-          <div>
-            <h3>
-              {name}
-              <span onClick={this.toggleDetails} className="btn btn-lg">
-                <FontAwesomeIcon icon={expandIcon} />
+          <div className="d-flex flex-column w-100">
```
Comment (sarah-clements) [inline ui/push-health/Metric.jsx]: It might be easier to read at a glance if you use the reactstrap layout elements Row, Col, etc.

File: ui/push-health/TestFailure.jsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      testName,
+      jobName,
+      jobId,
+      classification,
+      failureLine,
+      confidence,
+    } = failure;
+
+    return (
+      <div className="d-flex flex-column mt-2 mb-3 ml-2">
```
Comment (edmorley) [inline ui/push-health/TestFailure.jsx]: Could you switch this to reactstrap instead?

Comment (camd) [reply ui/push-health/TestFailure.jsx]: Yeah, good call.  Done.

File: ui/push-health/Metric.jsx (hunk 1 above)
Comment (camd) [reply ui/push-health/Metric.jsx]: Good call.  Done

**Comment by camd** [comment]
> In my mind our health score/test would be like:
> 7-10 - accept annotation (maybe some edge cases based on classification)
> <7 = make it easy to retrigger and collect more data
> 
> we could do this further and say:
> 1-4 = no confidence at all, or no idea what this is
> 5-7 = a good chance this is a [infra, leak, intermittent, regression, etc.], but we need more data
> 
> of course retriggering a job 2 more times so we have 3 data points will be a good first step towards marking it as intermittent/noproblem if 2/3 are green. There is an edge case here when a job has the opportunity for many frequent intermittent failures and run 1 fails on test_a.html, run 2 fails on test_b.html and run 3 fails on test_c.html - therefore it could look like perma fail- since we are doing test level failures instead of job/task level.
> 
> Of course, if we have health score in 8-10 range for a regression, then we might want to avoid showing a retrigger button until they fix the regression and push again.

This sounds good.  I think this would be best handled in the API logic and have that just report down to this UI as whichever state is appropriate.  So I'll copy these comments into the "data" doc we have.

**Comment by camd** [comment]
> In my mind our health score/test would be like:
> 7-10 - accept annotation (maybe some edge cases based on classification)
> <7 = make it easy to retrigger and collect more data
> 
> we could do this further and say:
> 1-4 = no confidence at all, or no idea what this is
> 5-7 = a good chance this is a [infra, leak, intermittent, regression, etc.], but we need more data
> 
> of course retriggering a job 2 more times so we have 3 data points will be a good first step towards marking it as intermittent/noproblem if 2/3 are green. There is an edge case here when a job has the opportunity for many frequent intermittent failures and run 1 fails on test_a.html, run 2 fails on test_b.html and run 3 fails on test_c.html - therefore it could look like perma fail- since we are doing test level failures instead of job/task level.
> 
> Of course, if we have health score in 8-10 range for a regression, then we might want to avoid showing a retrigger button until they fix the regression and push again.

This sounds good.  I think this would be best handled in the API logic and have that just report down to this UI as whichever state is appropriate.  So I'll copy these comments into the "data" doc we have.

**Comment by camd** [comment]
I believe I've addressed all the comments here, so I'm going to merge this PR as approved.  @edmorley if you have more things I should correct, I'll do them in a follow-up.  I have another PR I want to submit on top of this one.  :)
